text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,1238 @@
+```
+
+**Alternate:** But what if I manually edited `Cargo.toml` instead of `cargo add`?
+Here, we can shortcut some questions about version requirements because clap aligns on minor releases.
+```toml
+[package]
+name = "foo"
+version = "0.1.0"
+edition = "2024"
+rust-version = "auto"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+clap = { version = "5.10.30", features = ["derive"] }
+clap_complete = "5.10"  # <-- new
+```
+And away I go:
+```console
+$ cargo check
+Warning: adding clap_complete@5.10.40 because 5.11.0 requires Rust 1.93 while you are running 1.92.
+To use the clap_complete@5.11.0 with a compatible Rust version, run `rustup update && cargo update`.
+To force the use of clap_complete@5.11.0 independent of your toolchain, run `cargo update --ignore-rust-version`
+```
+But I am in a hurry and don't want to disrupt my flow.
+`clap_complete@5.10.40` is likely fine.
+I am running `clap@5.10.30` and that has been working for me.
+I might even run [`cargo deny`](https://crates.io/crates/cargo-deny) to see if there are known vulnerabilities.
+So I continue development.
+
+Later I run:
+```console
+$ cargo update
+Name          Current Latest Note
+============= ======= ====== ==================
+clap          5.10.30 5.11.0 requires Rust 1.93
+clap_complete 5.10.40 5.11.0 requires Rust 1.93
+note: To use the latest depednencies, run `rustup update && cargo update`.
+To force the use of the latest dependencies, independent of your toolchain, run `cargo update --ignore-rust-version`
+$ rustup update
+Downloading and install 1.94
+$ cargo update
+Updating clap v5.10.30 -> v5.11.0
+Updating clap_complete v5.10.40 -> v5.11.0
+```
+
+At this point, I want to publish
+```console
+$ cargo publish
+... crates.io error about missing fields
+$ $EDITOR `Cargo.toml`
+$ cargo publish
+Published foo 0.1.0
+```
+If I look on crates.io, the new 0.1.0 version shows up with a rust-version of 1.94
+without me having to manual update the field and
+relying on the `cargo publish`s verify step to verify the correctness of that MSRV.
+
+</details>
+
+### Extended "MSRV" with an application
+
+I am developing an application using a certified toolchain.
+I specify this toolchain using a `rust-toolchain.toml` file.
+
+Rust 1.94 is the latest but my certified toolchain is 1.92.
+
+<details><summary>Expand for step through of this workflow</summary>
+
+At some point, I start a project:
+```console
+$ cargo new foo
+$ cat foo/Cargo.toml
+```
+```toml
+[package]
+name = "foo"
+version = "0.1.0"
+edition = "2024"
+rust-version = "auto"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+```
+```console
+$ cargo add clap -F derive
+Adding clap 5.10.30
+warning: clap 5.11.0 exists but requires Rust 1.93 while you are running 1.92.
+To use the clap@5.11.0 with a compatible Rust version, run `rustup update && cargo add clap@5.10.0`.
```

Comment:
Running `rustup update` won't help if there's a toolchain file being used as this case mentions. It may be better not to mention rustup, since it's also possible that we're not using a rustup distribution.

Reply:
Note: this is more for the implementation, than the RFC.

We can detect rustup is being used and do so today in many cases to specialize the error messages.  As of today, I don't think we can detect the use of a `rust-toolchain.toml` file.

For myself, I think this message is worth it enough for when there isn't a toolchain file but we can figure those things out during the implementation.